e2e/suite_test.go

```diff
@@ -25,93 +24,83 @@ func Test_All(t *testing.T) {
-			} else {
-				tests = append(tests, testName)
-			}
-		}
-		t.Logf("Will run the following Agentbaker E2E tests: %s", strings.Join(tests, ", "))
-	} else {
-		t.Logf("Running all Agentbaker E2E tests...")
-	}
+	scenarioTable := scenario.InitScenarioTable(t, suiteConfig.scenariosToRun)
 
 	cloud, err := newAzureClient(suiteConfig.subscription)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if err := setupCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, suiteConfig.clusterName); err != nil {
+	if err := ensureResourceGroup(ctx, t, cloud, suiteConfig.resourceGroupName); err != nil {
 		t.Fatal(err)
 	}
 
-	subnetID, err := getClusterSubnetID(ctx, cloud, suiteConfig.location, suiteConfig.resourceGroupName, suiteConfig.clusterName)
+	clusters, err := listClusters(ctx, t, cloud, suiteConfig.resourceGroupName)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	kube, err := getClusterKubeClient(ctx, cloud, suiteConfig)
-	if err != nil {
-		t.Fatal(err)
-	}
+	for _, scenario := range scenarioTable {
+		scenario := scenario
+
+		chosenCluster := chooseCompatibleCluster(scenario, clusters)
+		chosenClusterExists := chosenCluster != nil
+		if !chosenClusterExists {
+			t.Logf("could not find test cluster able to run scenario %q, creating a new one...", scenario.Name)
+			newCluster := getBaseClusterModel(
+				fmt.Sprintf(testClusterNameTemplate, randomLowercaseString(r, 5)),
+				suiteConfig.location,
+			)
+			chosenCluster = &newCluster
+			scenario.ScenarioConfig.ClusterMutator(chosenCluster)
+		}
 
-	if err := ensureDebugDaemonset(ctx, kube, suiteConfig.resourceGroupName, suiteConfig.clusterName); err != nil {
-		t.Fatal(err)
-	}
+		if err := ensureCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, chosenCluster, !chosenClusterExists); err != nil {
+			t.Fatal(err)
+		}
+
+		clusterName := *chosenCluster.Name
+		t.Logf("chosen cluster name: %q", clusterName)
 
-	var clusterParams map[string]string
-	err = wait.PollImmediateWithContext(ctx, 15*time.Second, 5*time.Minute, func(ctx context.Context) (bool, error) {
-		params, err := extractClusterParameters(ctx, t, kube)
+		subnetID, err := getClusterSubnetID(ctx, cloud, suiteConfig.location, *chosenCluster.Properties.NodeResourceGroup, clusterName)
 		if err != nil {
-			t.Logf("error extracting cluster parameters: %q", err)
-			return false, nil
+			t.Fatal(err)
 		}
-		clusterParams = params
-		return true, nil
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
 
-	if err := createClusterParamsDir(); err != nil {
-		t.Fatal(err)
-	}
+		kube, err := getClusterKubeClient(ctx, cloud, suiteConfig.resourceGroupName, clusterName)
+		if err != nil {
+			t.Fatal(err)
+		}
 
-	t.Logf("dumping cluster parameters to local directory: %s", clusterParamsDir)
-	if err := dumpFileMapToDir(clusterParamsDir, clusterParams); err != nil {
-		t.Error("error dumping cluster parameters", err)
-	}
+		if err := ensureDebugDaemonset(ctx, kube); err != nil {
```

Comment:
should we extract this into a cluster setup function?

Reply:
yeees, was already thinking about that